packages/tangle/parser.go

```diff
@@ -248,6 +251,157 @@ type MessageFilter interface {
 	OnReject(callback func(msg *Message, err error, peer *peer.Peer))
 }
 
+/**
+Syntactical checks for a message
+*/
+type MessageSyntacticalValidator struct {
+	onAcceptCallback func(msg *Message, peer *peer.Peer)
+	onRejectCallback func(msg *Message, err error, peer *peer.Peer)
+
+	onAcceptCallbackMutex sync.RWMutex
+	onRejectCallbackMutex sync.RWMutex
+}
+
+func NewMessageSyntacticalValidator() *MessageSyntacticalValidator {
+	return &MessageSyntacticalValidator{}
+}
+
+/**
+Performs ths following syntactical checks:
+1. A Strong Parents Block must exist.
+2. Parents Block types cannot repeat.
+3. Parent count per block 0 <= x <= 8.
+4. Parents unique within block.
+5. Parents lexicographically sorted within block.
+6. ParentsBlockCount must match blocks actually parsed.
+7. ParentsCount within each block must match the parents actually parsed within the block.
+8. A Parent(s) repetition is only allowed when it occurs across Strong and Like parents
+8. Blocks should be ordered by type in ascending order
+**/
+func (f *MessageSyntacticalValidator) Filter(msg *Message, peer *peer.Peer) {
+	// A strong Block must exist
+	if len(msg.ParentsByType(StrongParentType)) == 0 {
+		f.onRejectCallback(msg, ErrNoStrongParents, peer)
+		return
+	}
+
+	// validate block count
+	if int(msg.parentsBlocksCount) != len(msg.parentsBlocks) {
+		f.onRejectCallback(msg, ErrBlockCountMismatch, peer)
+		return
+	}
+
+	if !f.validateUniqueness(msg, peer) {
+		return
+	}
+
+	// Parent count per block 0 <= x <= 8
+	for _, block := range msg.parentsBlocks {
```

Comment:
I think you could use `ForEachParentByType` and `ForEachParent` to make the code more readeable.

Reply:
using forEachParent now complicates the code

This is because it has a functional style that utilizes a consumer.
I want now to return from my upper-level function in the middle and this makes it hard w.o doing some tricks

besides as far as I understnad golang, the ethos is to keep things simple, and for loops are simple as it gets